### PR TITLE
chore(prerelease): do meaningful prerelease

### PR DESCRIPTION
**Pull request by louis-bompart**
## Proposed changes

While working, due to the misfire that occurred, the prerelease version is not really meaningful: the behaviour I wanted was 'the next release prefixed by -someNumber'. Now it is not, and if other mishaps happened, it could have been the same.

So to fix that, I changed how we define the beta version: we base strictly ourselves on two things:
 - the next 'gold/release' version.
 - the versions existing in the registry.

If we find no unprefixed prerelease for the next gold, then we create the -0 one, otherwise, we took the latest and 'bump it'

## Testing:

On d8c5c7a252fd5268b0bbedaec6d25a11cbd88662, run `npm run release:phase1 -w=@coveo/cli`.
Prior to the change, it was doing `3.0.0-1`. Now it is doing `2.0.0-1`.

### CODE REVIEW

File: scripts/releaseV2/phase1-bump-all-packages.mjs [hunk 1]
```diff
@@ -16,7 +16,8 @@ import angularChangelogConvention from 'conventional-changelog-angular';
 import {dirname, resolve, join} from 'path';
 import {fileURLToPath} from 'url';
 import retry from 'async-retry';
-import {gt, prerelease} from 'semver';
+import {gt, inc, prerelease} from 'semver';
+import {json as fetchNpm} from 'npm-registry-fetch';
```
Comment (louis-bompart) [inline scripts/releaseV2/phase1-bump-all-packages.mjs:20]: that the package npm cli is using to poke the registry. Don't mind if I do.

File: scripts/releaseV2/phase1-bump-all-packages.mjs [hunk 2]
```diff
@@ -180,3 +168,16 @@
   );
   return packageJson.private;
 }
+
+async function getNextBetaVersion(packageName, nextGoldVersion) {
+  const registryMeta = await fetchNpm(packageName);
+  const versions = Object.keys(registryMeta.versions);
+  const nextGoldMatcher = new RegExp(`${nextGoldVersion}-\d+`);
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply scripts/releaseV2/phase1-bump-all-packages.mjs:175]: Good bot.
```suggestion
  const nextGoldMatcher = new RegExp(`${nextGoldVersion}-\\d+`);
```

File: scripts/releaseV2/phase1-bump-all-packages.mjs [hunk 3]
```diff
@@ -100,7 +88,7 @@ const isPrerelease = process.env.IS_PRERELEASE === 'true';
     await writeChangelog(PATH, changelog);
   }
   const tagToPublish = isPrerelease ? 'beta' : 'latest';
-  await npmPublish('.', {tag: tagToPublish});
+  await npmPublish('.', {tag: tagToPublish, 'dry-run': true});
```
Comment (louis-bompart) [inline scripts/releaseV2/phase1-bump-all-packages.mjs:91]:
```suggestion
  await npmPublish('.', {tag: tagToPublish});
```

File: scripts/releaseV2/phase1-bump-all-packages.mjs [hunk 4]
```diff
@@ -180,3 +168,16 @@ function isPrivatePackage() {
...
+  const registryMeta = await fetchNpm(packageName);
+  const versions = Object.keys(registryMeta.versions);
+  const nextGoldMatcher = new RegExp(`${nextGoldVersion}-\\d+`);
+  const matchingPreReleasedVersions = versions
+    .filter((version) => nextGoldMatcher.test(version))
+    .sort();
+  if (matchingPreReleasedVersions.length === 0) {
+    return `${nextGoldVersion}-0`;
+  }
+  return inc(matchingPreReleasedVersions.pop(), 'prerelease');
```
Comment (y-lakhdar) [inline scripts/releaseV2/phase1-bump-all-packages.mjs:182]: nice

**Review by y-lakhdar — APPROVED** [review]

**Pull request merged**